base_export_async/models/delay_export.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2019 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+import json
+import operator
+import base64
+from dateutil.relativedelta import relativedelta
+
+from odoo import api, fields, models, _
+from odoo.addons.queue_job.job import job
+from odoo.addons.web.controllers.main import CSVExport, ExcelExport
+from odoo.exceptions import Warning
+
+_logger = logging.getLogger(__name__)
+
+
+class DelayExport(models.Model):
+
+    _name = 'delay.export'
+    _description = 'Allow to delay the export'
+
+    @api.model
+    def delay_export(self, data):
+        params = json.loads(data.get('data'))
+        if not self.env.user.email:
+            raise Warning(_("You must set an email address to your user."))
+        self.with_delay().export(params)
+
+    @api.model
+    @job
+    def export(self, params):
+        export_format = params.get('format')
+        raw_data = export_format != 'csv'
+
+        model_name, fields_name, ids, domain, import_compat, context = \
+            operator.itemgetter('model', 'fields', 'ids',
+                                'domain', 'import_compat', 'context')(params)
+        user = self.env['res.users'].browse([context.get('uid')])
+        if not user or not user.email:
+            raise Warning(_("The user doesn't have an email address."))
+
+        model = self.env[model_name].with_context(
+            import_compat=import_compat, **context)
+        records = model.browse(ids) or model.search(
+            domain, offset=0, limit=False, order=False)
+
+        if not model._is_an_ordinary_table():
+            fields_name = [field for field in fields_name
+                           if field['name'] != 'id']
+
+        field_names = [f['name'] for f in fields_name]
+        import_data = records.export_data(
+            field_names, raw_data).get('datas', [])
+
+        if import_compat:
+            columns_headers = field_names
+        else:
+            columns_headers = [val['label'].strip() for val in fields_name]
+
+        if export_format == 'csv':
+            csv = CSVExport()
+            result = csv.from_data(columns_headers, import_data)
+        else:
+            xls = ExcelExport()
+            result = xls.from_data(columns_headers, import_data)
+
+        name = "{}.{}".format(model_name, export_format)
+        attachment = self.env['ir.attachment'].create({
+            'name': name,
+            'datas': base64.b64encode(result),
+            'datas_fname': name,
+            'type': 'binary',
+            'to_delete': True,
+        })
+
+        url = "{}/web/content/ir.attachment/{}/datas/{}?download=true".format(
+            self.env['ir.config_parameter'].sudo().get_param('web.base.url'),
+            attachment.id,
+            attachment.name,
+        )
```

Comment:
Is there a need to secure access to these attachments, so only the user who requested the export can read them?

Reply:
Maybe we need to secure it yes but the email that contains the link is already secured and send to the user only. I'm open to suggestion.

Reply:
We are going to link the attachment to a record of the delay.export. We are going to create a ir.rule on the delay.export so that only the user himself can read it. Therefore the attachment will be protected.